Feature/mpc specific simulation setup

Beomyeong opened the pull request:
Re-work for whole body MPC specific simulation environment. 
This branch works for two things. 

1. MPC specific simulation setup to be run independent with regular walking controller
2. `WholeBodyControllerCoreThread` is added. 
 
Objectives of this branch  are
- ControllerThread populate mpcControllerOutput to the wholebodycontrollercoreThread(WBCC Thread)
- MPCControllerOutput is called in WBCC Thread
- WBCC will be computed in WBCC Thread with MPCControllerOutput
- Populate the desiredJointOutput to the actuator in the WBCC Thread

It seems like robot shaking with the same tuning weighs. 
Doubt is that computing MPC in controllerThread and populating it in WBCC Thread. 
I hope calculating WBCC with MPC controllerOutput in WBCC thread solves this issue.

Beomyeong committed:
Re-starting to make MPC specific simulation setups.
- Adding SCS2AvatarMPCSimulation
- Adding SCS2AvatarMPCSimulationFactory
- Editing MPC specific classes in NadiaMPCSimulation and NadiaMPCSimulationFactory

Beomyeong committed:
- Adding AvatarMPCControllerThread and MPCHighLevelHumanoidControllerFactory.
- Changing the class in NadiaMPCSimulationFactory, SCS2AvatarMPCSimulation and SCS2AvatarMPCSimulationFactory.

Beomyeong committed:
- Adding MPCHumanoidHighLevelControlManager
- Modifying called class name.

Beomyeong committed:
Just changing variable name

Beomyeong committed:
- Adding MPCHighLevelControllerState and changing mentioned class name

Beomyeong committed:
- Adding WBCC thread and wbcc highlevel manager

Beomyeong committed:
- Adding MPCControllerTask

Beomyeong committed:
- Adding MPC estimatorthreadFactory and task
- Adding MPC StepgeneratorThread and task

Beomyeong committed:
- Adding wholebodycontrollercoretask.
- wbcc thread is added to the controller. But that thread is still empty, do nothing.

Beomyeong committed:
- Adding HumanoidRobotMPCContextData and its factory

Beomyeong committed:
- Adding MPCBarrierScheduledRobotController and MPCHumanoidRobotControlTask.
- Changing the parents class for MPCControllerTaks, MPCEstimatorTaks, MPCStepGeneratorTask and MPCWholeBodyControllerCoreTask to MPCHumanoidRobotControlTask.
- boolean of wholebodycontrollercoreRan is added to HumanoidRobotMPCContextData.

Beomyeong committed:
- WholeBodyControllerCoreThread is triggered so that the HumanoidWholeBodyControllerCoreManager is called in every tick.
- HWBCCM is still empty, so it do nothing.

Beomyeong committed:
- Adding MPCCrossRobotCommandResolver which has CrossRobotCommandResolver as parent class.
- MPCControllerTask, MPCWholeBodyControllerCoreTask is connected to the MPCCrossRobotCommandResolver. But the MPCEstimatorTask is still connected to the CrossRobotCommandResolver. Because, its updates to Master and to local has not been changes.

Beomyeong committed:
- HumanoidRobotMPCContext includes the mpcControlOutputDataList.
- This data type will contain the result of the mpcController and will be delivered to MPCWholeBodyControllerCoreThread from MPCControllerThread

Beomyeong committed:
Including desiredMPCControllerOutputDataHolder into constructors of each tread.

Beomyeong committed:
Merge branch 'develop' into feature/MPC-specific-simulationsetup

Beomyeong committed:
- MPCControllerOutput is saved into mpcControllerOutputDataHolder in HumanoidRobotMPCContextData.
- In HumanoidWholeBodyControllerCoreManager, the desiredMPCControllerOutput is called and then saved into `desiredJointOutput`which is populated into actuator.

rjgriffin42 commented on ihmc-avatar-interfaces/src/main/java/us/ihmc/avatar/AvatarMPCControllerThread.java:
```diff
@@ -0,0 +1,409 @@
+package us.ihmc.avatar;
```
Is there a reason that this is in open robotics and not in the crocoddyl repo?

Beomyeong replied:
It doesn't have much meaning. I just put this class to the place where AvatarControllerThread is. 
Do you think moving this class into crocoddyl repo is better?

Beomyeong replied:
Got it. I saw the below comment now,
I agree with you. 
I would move all the mpc specific classes to the crocoddyl repo.
 < we talked in person >

rjgriffin42 commented on ihmc-avatar-interfaces/src/main/java/us/ihmc/avatar/AvatarMPCControllerThread.java:
```diff
@@ -0,0 +1,409 @@
+
+public class AvatarMPCControllerThread implements AvatarControllerThreadInterface
+{
+   private static final boolean CREATE_DYNAMICALLY_CONSISTENT_NULLSPACE_EVALUATOR = false;
```
I think this could be removed

rjgriffin42 commented on ihmc-common-walking-control-modules/src/main/java/us/ihmc/commonWalkingControlModules/barrierScheduler/context/HumanoidRobotMPCContextData.java:
```diff
@@ -0,0 +1,102 @@
+package us.ihmc.commonWalkingControlModules.barrierScheduler.context;
```
So when looking at this code, it doesn't look like any of the member variables are used. What's the specific reason for this class to exist, as opposed to just using the regular HumanoidRobotContextData?

Beomyeong replied:
The `HumanoidRobotMPCContextData` includes additional dataholders than normal `HumanoidRobotMPCContextData`. (I just call  the data which included in the HumanoidRobotContextData as dataholder). 
e.g. mpcDesiredOutputData is included in `HumanoidRobotMPCContextData`.  WholeBodyControllerCoreCommand will be included later. 
While writing this, something came to mind. I guess that when adding new variables into `HumanoidRobotContextData`, every variables should be initialized. And this causes changes to all the controllers. But, I might be wrong. Let me check this, and then I'll update it.

Beomyeong replied:
Yes. If I add new variables to the regular HumanoidRobotContextData, I should change the regular classes when construct the HumanoidRobotContextData. 
The HumanoidRobotMPCContextData could be replace with HumanoidRobotContextData by including the variables. But for now, I would like to use `HumanoidRobotMPCContextData`, after we prepare to move to the MPCController, I would work to `HumanoidRobotContextData` will replace MPC One.

rjgriffin42 commented on ihmc-common-walking-control-modules/src/main/java/us/ihmc/commonWalkingControlModules/barrierScheduler/context/HumanoidRobotMPCContextDataFactory.java:
```diff
@@ -0,0 +1,30 @@
+package us.ihmc.commonWalkingControlModules.barrierScheduler.context;
```
If the context data class isn't needed, this factory isn't needed.

rjgriffin42 commented on ihmc-common-walking-control-modules/src/main/java/us/ihmc/commonWalkingControlModules/controllerCore/command/MPCCrossRobotCommandResolver.java:
```diff
@@ -0,0 +1,51 @@
+import us.ihmc.robotModels.RigidBodyHashCodeResolver;
+import us.ihmc.sensorProcessing.frames.ReferenceFrameHashCodeResolver;
+
+public class MPCCrossRobotCommandResolver extends CrossRobotCommandResolver
```
If the context data class isn't needed, this command resolver isn't needed. Etiher way, I'd consider pushing these methods into the parent class, as opposed to to using inheritence.

Beomyeong replied:
The content of resolverHumanoidRobotContextDataController and resolveHumanoidRootContextDataControllerCore in both HumanoidRobotMPCContextData and HumanoidRobotContextData are different. 
While some methods are identical e.g. resolveHumanoidRobotContextDataEstimator

rjgriffin42 commented on ihmc-avatar-interfaces/src/main/java/us/ihmc/avatar/MPCStepGeneratorTask.java:
```diff
@@ -0,0 +1,22 @@
+package us.ihmc.avatar;
```
Is there any difference between this and the regular one? I'd  use the regular one, if there is not.

Beomyeong replied:
In the method of updateMasterContext, the tyoe of HumaonidRobotMPCContextData is used as input.

rjgriffin42 commented on ihmc-avatar-interfaces/src/main/java/us/ihmc/avatar/MPCEstimatorTask.java:
```diff
@@ -0,0 +1,99 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class MPCEstimatorTask extends MPCHumanoidRobotControlTask
```
Is there any difference between this and the regular one? I'd use the regular one if there is not.

Beomyeong replied:
The updateMasterContext and updateLocalContext use different type of input ("HumaonidRobotMPCContextData")

rjgriffin42 commented on ihmc-avatar-interfaces/src/main/java/us/ihmc/avatar/MPCWholeBodyControllerCoreTask.java:
```diff
@@ -0,0 +1,86 @@
+   {
+      wholeBodyControllerCoreResolver.resolveHumanoidRobotContextDataScheduler(masterContext, (HumanoidRobotMPCContextData) wholeBodyControllerCoreThread.getHumanoidRobotContextData());
+      wholeBodyControllerCoreResolver.resolveHumanoidRobotContextDataEstimator(masterContext, (HumanoidRobotMPCContextData) wholeBodyControllerCoreThread.getHumanoidRobotContextData());
+      wholeBodyControllerCoreResolver.resolveHumanoidRobotContextDataControllerCore(masterContext, (HumanoidRobotMPCContextData) wholeBodyControllerCoreThread.getHumanoidRobotContextData());
```
It may be preferred to do the casting as an additional line, and hten pass it as an argument, so that you don't cast three times.

rjgriffin42 commented on ihmc-avatar-interfaces/src/main/java/us/ihmc/avatar/AvatarMPCWholeBodyControllerCoreThread.java:
```diff
@@ -0,0 +1,291 @@
+   private final YoGraphicsListRegistry yoGraphicsListRegistry = new YoGraphicsListRegistry();
+   private final List<Supplier<YoGraphicDefinition>> scs2YoGraphicHolders = new ArrayList<>();
+   private ModularRobotController wholeBodyControllerCoreCalculator = new ModularRobotController("WBCC");
+   private final ExecutionTimer wholeBodyControllerCoreThreadTimer;
```
this is never used.

Beomyeong replied:
I forgot to remove it when I checked the computation time.

rjgriffin42 commented on ihmc-avatar-interfaces/src/main/java/us/ihmc/avatar/AvatarMPCWholeBodyControllerCoreThread.java:
```diff
@@ -0,0 +1,291 @@
+   private final YoBoolean runWholeBodyControllerCore = new YoBoolean("RunWholeBodyControllerCore", registry);
+   private final YoGraphicsListRegistry yoGraphicsListRegistry = new YoGraphicsListRegistry();
+   private final List<Supplier<YoGraphicDefinition>> scs2YoGraphicHolders = new ArrayList<>();
+   private ModularRobotController wholeBodyControllerCoreCalculator = new ModularRobotController("WBCC");
```
maybe name "controller" instead of "calculator", so it's more obvious it's a modular robot controller.

Beomyeong replied:
Changed

rjgriffin42 commented on ihmc-common-walking-control-modules/src/main/java/us/ihmc/commonWalkingControlModules/highLevelHumanoidControl/HumanoidWholeBodyControllerCoreManager.java:
```diff
@@ -0,0 +1,169 @@
+         // So, the output of other highLevelStates should be called in here to be delivered to copyJointDesiredsToJoints
+         yoDesiredMPCControllerOutputDataHolder.overwriteWith(desiredMPCControllerOutput);
+         controllerCore.submitControllerCoreCommand(controllerCoreCommand);
+         //         controllerCore.compute();
```
how does this work if you're not calling compute?

Beomyeong replied:
Currently, only the MPCController Output which comes from the controllerThread is used to populate to the actuator. 
`mpcController.getMPCSolution()` in MPCControllerState are saved to desiredMPCCOntrollerOutput in `MPCHumanoidHighLevelControllerManager.copyJointDesiredsToJoints`. And, this is delivered to wholebodycontrollercoreThread by HumanoidRobotMPCContextData. 
Finally, desiredMPCControllerCoreOutput is saved to lowLevelControllerOutput in the copyJointDesiredsToJoints in `HumanoidWholeBodyControllerCoreManager`

rjgriffin42 commented on ihmc-common-walking-control-modules/src/main/java/us/ihmc/commonWalkingControlModules/highLevelHumanoidControl/MPCHumanoidHighLevelControllerManager.java:
```diff
@@ -0,0 +1,427 @@
+package us.ihmc.commonWalkingControlModules.highLevelHumanoidControl;
```
I'm not sure this needs to exist.  Is there anything currently different from the regular one?

Beomyeong replied:
There are a few differences. 
1. MPC one has input value of desiredMPCControllerOutput
2. In the copyJointDesiredToJoints, MPC one populates the mpcControllerOutput, instead regular one populate the wholebodyControllerCoreOutput.

The second is key point, and first one comes from second one

rjgriffin42 commented:
I think this is a step in the right direction, but with a few areas for improvement:
- I'm not sure I see how it actually works?
- It looks like there were a lot of classes that were copy-pasted that didn't need to be, as they're still functionally the same.
- I would prefer it if the classes that are specific to the MPC controller were in the MPC repo, as opposed to in open source.

Beomyeong committed:
Moving all the classes from open-robotics to crocoddyl repo

Beomyeong committed:
- Adding ControllerCoreCommandDataHolder
- Let the wholeBodyControllerCore compute be called in WBCC Thread.
- The controllerCoreCommand which includes the MPCControllerOutput is delivered to WBCC thread from ControllerThread.
- StateError Vector in MPC with WBCC is still updated in ControllerThread.

rjgriffin42 commented on ihmc-whole-body-controller/src/main/java/us/ihmc/wholeBodyController/WholeBodyControllerParameters.java:
```diff
@@ -1,62 +1,63 @@
    }
 
-   public StateEstimatorParameters getStateEstimatorParameters();
+   double getWBCCDT();
```
I think this name should be spelled out. `getWholeBodyControllerCoreDT()`

rjgriffin42 commented on ihmc-common-walking-control-modules/src/main/java/us/ihmc/commonWalkingControlModules/highLevelHumanoidControl/highLevelStates/HighLevelControllerState.java:
```diff
@@ -72,6 +73,11 @@ public RootJointDesiredConfigurationDataReadOnly getOutputForRootJoint()
    {
       return null;
    }
+   public ControllerCoreCommand getControllerCoreCommand()
```
space above this?

rjgriffin42 commented on ihmc-common-walking-control-modules/src/main/java/us/ihmc/commonWalkingControlModules/controllerCore/command/ControllerCoreCommandDataHolder.java:
```diff
@@ -0,0 +1,203 @@
+      hasVirtualModelControlCommandList = true;
+      hasControllerCoreMode = true;
+      hasFeedbackControlCommandList = true;
+      hasLowLevelOneDoFJointDesiredDataHolder = true;
```
this should be getting the value from the other object, not setting them all to be true. For example, if other.getInverseDynamicsCommandList() is empty, then hasInverseDynamicsCommandList should be false.

rjgriffin42 commented on ihmc-common-walking-control-modules/src/main/java/us/ihmc/commonWalkingControlModules/controllerCore/command/ControllerCoreCommandDataHolder.java:
```diff
@@ -0,0 +1,203 @@
+      }
+      else if (obj instanceof ControllerCoreCommandDataHolder other)
+      {
+         if (!inverseKinematicsCommandList.equals(other.inverseKinematicsCommandList))
```
You should run through the "has" cehcks, as well as the reinitialize check

Beomyeong committed:
Merge branch 'develop' into feature/MPC-specific-simulationsetup

Beomyeong committed:
MPCControllerSolutionDataHolder is added.
SolutionFloatingBasePose, --baseLinearVelocity, --AngularVelocity, and lowLevelMPCControllerOutputs(Joint position, velocities, and torques) are saved into this dataHolder.
** From the visual graphic, contact wrench seems like bigger than before. Need to check this. However, the MPCChestSimulation works well, too.

Beomyeong committed:
All the method related to calculate MPCState error moved from controller Thread to wbcc Thread.
But, there are still errors.

Beomyeong committed:
Merge branch 'develop' into feature/MPC-specific-simulationsetup

Beomyeong committed:
Tweaked small portions. Change the value name, and add boolean set

Beomyeong committed:
Merge branch 'develop' into feature/MPC-specific-simulationsetup

james-p-foster commented on atlas/src/main/java/us/ihmc/atlas/AtlasRobotModel.java:
```diff
@@ -466,6 +454,12 @@ public double getControllerDT()
+   @Override
+   public double getWholeBodyControllerCoreDT()
+   {
+      return 0.0;
```
TODO comment on what change should eventually happen here please

Beomyeong replied:
Done

james-p-foster commented on atlas/src/main/java/us/ihmc/atlas/AtlasRobotModel.java:
Not too fussed, but I'd personally advocate for reverting the formatting changes done here. They aren't relevant to the PR and we don't tend to change `AtlasRobotModel` that much

james-p-foster commented on ihmc-common-walking-control-modules/src/main/java/us/ihmc/commonWalkingControlModules/highLevelHumanoidControl/highLevelStates/HighLevelControllerState.java:
```diff
@@ -73,6 +77,50 @@ public RootJointDesiredConfigurationDataReadOnly getOutputForRootJoint()
+   {
+      return 0;
+   }
+   public double getMpcCostCommandWeight()
```
Capitalise MPC:
```suggestion
   public double getMPCCostCommandWeight()
```

Beomyeong replied:
Done

james-p-foster commented on ihmc-whole-body-controller/src/main/java/us/ihmc/wholeBodyController/WholeBodyControllerParameters.java:
```diff
@@ -1,62 +1,63 @@
    /**
     * Get the parameter XML file for the controller.
-    *
+    * <p>
```
Should have a matching `</p>` after the following paragraph

Beomyeong replied:
Done

james-p-foster commented on ihmc-whole-body-controller/src/main/java/us/ihmc/wholeBodyController/WholeBodyControllerParameters.java:
Why have all the access modifiers been removed? We should add them back in

Beomyeong replied:
Yeah. Its' funny. I'm not remember exactly, but, some other day, that modifiers made build errors, so I deleted everything.  But now, there is no issue related to those modifiers. So, I rolled back to original ones

james-p-foster committed:
Merge branch 'develop' into feature/MPC-specific-simulationsetup

james-p-foster committed:
Merge remote-tracking branch 'origin/develop' into feature/MPC-specific-simulationsetup

james-p-foster committed:
Merge remote-tracking branch 'origin/feature/MPC-specific-simulationsetup' into feature/MPC-specific-simulationsetup

Beomyeong committed:
Merge remote-tracking branch 'origin/feature/MPC-specific-simulationsetup' into feature/MPC-specific-simulationsetup

Beomyeong committed:
Roll back useless changes in AtlasRobotModel. And TODO is added to getWholeBodyControllerCoreDT method.

Beomyeong committed:
Changing in the name of method.
Revive the modifiers had been deleted.

rjgriffin42 commented on ihmc-common-walking-control-modules/src/main/java/us/ihmc/commonWalkingControlModules/controllerCore/command/ControllerCoreCommandDataHolder.java:
```diff
@@ -0,0 +1,212 @@
+            return false;
+         if (!controllerCoreMode.equals(other.controllerCoreMode))
+            return false;
+         return lowLevelOneDoFJointDesiredDataHolder.equals(other.lowLevelOneDoFJointDesiredDataHolder);
```
I am not convinced this is completely correct. Specifically, if, for example, `hasFeedbackControlCommandList` is false, I don't care if the values contained in the two `feedbackControlCommandList` objects are equivalent. So I'm sure this will always return true negatives, but I think it may also return false negatives by being overly restrictive.

Beomyeong replied:
This is done to include the comparing `hasFeedbackControlCommandList` is equal in the if statement for comparing dataList.

rjgriffin42 commented on ihmc-common-walking-control-modules/src/main/java/us/ihmc/commonWalkingControlModules/highLevelHumanoidControl/highLevelStates/HighLevelControllerState.java:
```diff
@@ -73,6 +77,50 @@ public RootJointDesiredConfigurationDataReadOnly getOutputForRootJoint()
+   public double getMPCCostCommandWeight()
+   {
+      return 0.0;
+   }
```
I don't think we want to add most of these fields. I don't think theyr'e used, if I remember correctly, and I know we don't want them here.

stefanfasano replied:
Perhaps we could make a MPCHighLevelControllerState that extends HighLevelControllerState and adds some of these fields?

james-p-foster replied:
That's what Beomyeong and I are gonna pair on today. Unfortunately it does start a cascade of duplicating other things, but after discussing with Beomyeong and Robert, a duplicate setup is probably the best way of doing things here

Beomyeong replied:
Yeap. Now we are on the same page

stefanfasano commented on ihmc-common-walking-control-modules/src/main/java/us/ihmc/commonWalkingControlModules/controllerCore/command/ControllerCoreCommandDataHolder.java:
```diff
@@ -0,0 +1,212 @@
+      return lowLevelOneDoFJointDesiredDataHolder;
+   }
+
+   public boolean equals(Object obj)
```
Rather than passing in a generic Object, can you pass in an instance of ControllerCoreCommandDataHolder? That would remove the "else" condition in this method. Is there a scenario where you would use this method and pass in something that isn't a ControllerCoreCommandDataHolder?

Beomyeong replied:
I agree with you. However, to maintain consistency with other data holder classes, I would leave it as it is. Other data holders are also using a generic objects directly in the equals method.

stefanfasano commented on ihmc-common-walking-control-modules/src/main/java/us/ihmc/commonWalkingControlModules/controllerCore/command/ControllerCoreCommandDataHolder.java:
What is the overall purpose of this class, and how is it fundamentally different than ControllerCoreCommand? Can you write a brief documentation at the top of this class explaining this purpose?

Beomyeong replied:
Done.

stefanfasano replied:
I'm not seeing a class description, but maybe it just isn't showing up on github. Other than that the PR looks good

Beomyeong committed:
Merge branch 'develop' into feature/MPC-specific-simulationsetup

Beomyeong committed:
Use casting in MPCHumanoidHighLevelControllerManager, which allows removing polluting MPC methods in HighLevelControllerState

Beomyeong committed:
Merge branch 'develop' into feature/MPC-specific-simulationsetup

Beomyeong committed:
Merge branch 'develop' into feature/MPC-specific-simulationsetup

Beomyeong committed:
Merge branch 'develop' into feature/MPC-specific-simulationsetup
